keps/sig-node/kubelet-node-endpoint.md

```diff
@@ -0,0 +1,89 @@
+---
+title: Kubelet Node Endpoint
+authors:
+  - "clamoriniere"
+owning-sig: sig-node
+participating-sigs: []
+reviewers: []
+approvers: []
+creation-date: 2019-05-02
+status: implementable
+---
+
+# Kubelet Node Endpoint
+
+## Motivation
+
+The Kubelet already proposes a set of endpoints (`/pods`, and soon `/metrics/resource/v1*`) that offer useful information for "local" node monitoring. In addition to those endpoints, we would like to introduce a new `/node` endpoint for displaying the node resource information. It will allow daemonset-based monitoring solutions to retrieve node information locally (on the same host) like "labels, annotations" instead of using a call to the API server, which will solve some scalability issues in big clusters.
+
+![kubelet-node-endpoint](https://i.ibb.co/khVg8rd/Untitled-Diagram.png)
+
+## Proposal
+
+Kubelet already caches and sync its corresponding `Node` information from the API-server. The proposed solution opens a new endpoint "/node" that returns the `k8s.io/api/pkg/core/v1.Node` JSON representation of the Kubelet's Node. Only the Kubelet's Node resource is return.
+
+For consistency the endpoint path "/node" looks similar to the "/pods" endpoint (no API versioning).
+
+## Use Cases
+
+* **Pod metrics and logs tagging with Node labels or annotation**: Monitoring workload solutions based on the kubelet API ("/pods" and "/metrics") used to decorate Pod metrics with additional context information; Information like the "Availability zone", "Provider", "Server type" etc are very useful for issue investigation.
+The current workaround is to retrieve that information from the API-Server, that generates a consequent number of requests to the API-Server on a large kubernetes cluster.
+* **Container images monitoring on Node**: Knowing which Images and Disk usage that it represents on a specific node can be useful for Node monitoring.
+* **Node conditions**: Node's health information is available in the “status.Conditions” list, like DiskPressure, MemoryPressure, PIDPressure… All those conditions are useful for triggering Alert monitoring and used “locally” for correlation with Pod misbehavior.
+* **Node Info and characteristic**: 
+   - Knowing the container runtime can be used to configure monitoring checks that interact directly with the container runtime.
```

Comment:
What kind of information are you trying to get through this?  It might be nice to have a mechanism for node agents to interact with the CRI, as this provides consistent information across CRI implementations.  But if you are trying to get runtime-specific information, then that wouldn't work.

Reply:
CRI define API to retrieve Container stats metrics, but unfortunately, the CRI didn't specify a monitoring|metrics API for the runtime itself: https://github.com/kubernetes/kubernetes/blob/242a97307b34076d5d8f5bbeb154fa4d97c9ef1d/docs/devel/container-runtime-interface.md#cri-1.5-known-issues

Currently, to retrieve container runtime metrics, you need to guess which runtime is used by looking at the Kubelet configuration or asking the user to provide the information.

Reply:
By "container runtime metrics", are you looking for metrics about containers, or metrics about the container runtime itself?

Reply:
container runtime itself

Reply:
Does the kubelet configuration actually tell you which runtime is being used?  IIRC, it just tells you which socket the container runtime is serving on...